File: kdtree.cc

#include <vector>
#include <cassert>
#include <iostream>
#include <functional>
#include "median.cc"
using namespace std;


template <typename T>
ostream& operator<<(ostream& os, const vector<T>& vec) {
    os << "[";
    for (size_t i = 0; i < vec.size(); ++i) {
        os << vec[i];
        if (i < vec.size() - 1) {
            os << ", ";
        }
    }
    os << "]";
    return os;
}

istream& operator>>(istream& is, vector<int>& vec) {
    string x;
    vec = vector<int>();
    while (is >> x && x != ".") {
        vec.push_back(stoi(x));
    }
    return is;
}


template <typename T>
class node {
    T val;
    node<T> *left;
    node<T> *right;
    int discr;
    int dim;
    int depth;
    node(int depth) : depth(depth), left(NULL), right(NULL) {
        k = depth % dim;
    }
    node(T val, int depth) : val(val), depth(depth), left(NULL), right(NULL) {
        dim = val.size();
        k = depth % dim;
    }
    ~node() {
        delete left;
        delete right;
    }
    void print(string prefix_root="", string prefix_left="", string prefix_right="") {
        string bar_space = "|" + string(dim*4-1, ' ');
        string space = string(dim*4, ' ');
        string line = string(dim*4 - 2, '-');
        if (left != NULL) left->print(prefix_left + " ." + line, prefix_left + space, prefix_left + bar_space);
        cout << prefix_root << val;
        cout << endl;
        if (right != NULL) right->print(prefix_right + " '" + line, prefix_right + bar_space, prefix_right + space);
    }
};


template <typename T>
class kdtree {

    node<T> *root;
    int dim;

    node<T>* insert_list(node<T> *p, const vector<vector<T>> &v) {
        if (v.size() == 0) {
            return NULL;
        }
        int k = p->k;
        vector<T> keys(v.size());
        for (int i = 0; i < v.size(); i++) {
            keys[i] = v[i][k];
        }
        quickselect(keys, keys.size()/2);
        
    }


    void insert(node<T> *p, const vector<T> &val) {
        if (val < p) {
            if (p->left == NULL) {
                p->left = new node<T>(val, p->depth+1);
            }
            else {
                insert(p->left, val);
            }
        }
        else {
            if (p->right == NULL) {
                p->right = new node<T>(val, p->depth+1);
            }
            else {
                insert(p->right, val);
            }
        }
    }

public:
    kdtree(int dim) : root(NULL), dim(dim) {
    }
    ~kdtree() {
        delete root;
    }
    kdtree(const vector<vector<T>> &v) : root(NULL), dim(v[0].size()) {
        root = insert_list(v);
    }
    kdtree(const kdtree &t) {
        this = t;
    }
    kdtree<T> &operator=(const kdtree &t) {
        return *this;
    }

    void insert(const vector<T> &val) {
        assert(val.size() == dim);
        if (root == NULL) {
            root = new node<T>(val, 0);
        }
        else {
            insert(root, val);
        }
    }

    vector<T> as_vector() {
        vector<T> v;
        if (root == NULL) return v;
        vector<node<T>*> stack;
        stack.push_back(root);
        while (!stack.empty()) {
            node<T> *p = stack.back();
            stack.pop_back();
            v.push_back(p->val);
            if (p->left != NULL) stack.push_back(p->left);
            if (p->right != NULL) stack.push_back(p->right);
        }
        return v;
    }

    void calculate_depth(node<T>* p, int depth, int& max_depth) {
        if (p == NULL) return;
        if (depth > max_depth) max_depth = depth;
        calculate_depth(p->left, depth+1, max_depth);
        calculate_depth(p->right, depth+1, max_depth);
    };

    int getDepth() {
        int max_depth = 0;
        calculate_depth(root, 1, max_depth);
        return max_depth;
    }

    void print() {
        cout << "kdtree" << endl;
        int depth = getDepth();
        cout << depth << endl;
        for (int i = 0; i < depth; i++) {
            cout << "(" << i%dim << ")" << string(dim*4-3, ' ');
        }
        cout << endl;

        if (root == NULL) return;
        root->print();
    }

};

int main(int argc, char *argv[]) {
    vector<vector<int>> V(0);
    vector<int> v;
    while (cin >> v && v.size() > 0) {
        V.push_back(v);
        cout << "v: " << v << endl;
    }
    for (int i = 0; i < V.size(); ++i) {
        cout << "V[" << i << "]: " << V[i] << endl;
    }
    kdtree<int> t(V);
    t.print();
    return 0;
}   

File: kdtree.py
import random
from typing import List, Optional, Union, Set

def format_num(num:int|float):
    if isinstance(num, int):
        return str(num)
    else:
        s = f'{num:.2f}'
        if s.startswith('0.'):
            s = s.removeprefix('0')
        return s

def format_vec(vec:list[int|float]):
    return '[' + ', '.join(format_num(x) for x in vec) + ']'

class RKDNode:
    def __init__(self, key: List[float], discr: int, left=None, right=None):
        self.key = key
        self.discr = discr
        self.left: Optional[RKDNode] = left
        self.right: Optional[RKDNode] = right
        self.size = 1 + (left.size if left else 0) + (right.size if right else 0)
        self.dim = len(key)

    def print(self, prefix_root="", prefix_left="", prefix_right=""):
        indent = self.dim * 5 + 4
        bar_space = " |" + " " * (indent - 2)
        space = " " * (indent)
        line = "-" * (indent - 2)
        if self.left:
            self.left.print(prefix_left + " ." + line, prefix_left + space, prefix_left + bar_space)
        print(f"{prefix_root}({self.discr}) {format_vec(self.key)}")
        if self.right:
            self.right.print(prefix_right + " '" + line, prefix_right + bar_space, prefix_right + space)



def update_size(node: Optional[RKDNode]):
    if node:
        node.size = 1 + (node.left.size if node.left else 0) + (node.right.size if node.right else 0)

def split(T: Optional[RKDNode], x: List[float], i: int) -> tuple[Optional[RKDNode], Optional[RKDNode]]:
    if not T:
        return None, None

    j = T.discr
    if i == j:
        if x[i] < T.key[i]:
            L, R = split(T.left, x, i)
            T.left = R
            update_size(T)
            return L, T
        else:
            L, R = split(T.right, x, i)
            T.right = L
            update_size(T)
            return T, R
    else:
        L1, R1 = split(T.left, x, i)
        L2, R2 = split(T.right, x, i)
        if x[i] < T.key[i]:
            left = join(L1, L2, j)
            T.left, T.right = R1, R2
            update_size(T)
            return left, T
        else:
            right = join(R1, R2, j)
            T.left, T.right = L1, L2
            update_size(T)
            return T, right

def join(U: Optional[RKDNode], V: Optional[RKDNode], i: int) -> Optional[RKDNode]:
    if not U:
        return V
    if not V:
        return U

    m, n = U.size, V.size
    if random.randint(0, m + n - 1) < m:
        j = U.discr
        if i == j:
            U.right = join(U.right, V, i)
            update_size(U)
            return U
        else:
            L, R = split(V, U.key, j)
            U.left = join(U.left, L, i)
            U.right = join(U.right, R, i)
            update_size(U)
            return U
    else:
        j = V.discr
        if i == j:
            V.left = join(U, V.left, i)
            update_size(V)
            return V
        else:
            L, R = split(U, V.key, j)
            V.left = join(L, V.left, i)
            V.right = join(R, V.right, i)
            update_size(V)
            return V

def insert_at_root(T: Optional[RKDNode], x: List[float]) -> RKDNode:
    i = random.randint(0, len(x) - 1)
    L, R = split(T, x, i)
    new_node = RKDNode(key=x, discr=i, left=L, right=R)
    update_size(new_node)
    return new_node

def insert(T: Optional[RKDNode], x: List[float]) -> RKDNode:
    n = T.size if T else 0
    if random.randint(0, n) == n:
        return insert_at_root(T, x)
    else:
        assert T is not None
        i = T.discr
        if x[i] < T.key[i]:
            T.left = insert(T.left, x)
        else:
            T.right = insert(T.right, x)
        update_size(T)
        return T

def delete_root(T: Optional[RKDNode]) -> Optional[RKDNode]:
    if not T:
        return None
    return join(T.left, T.right, T.discr)

def delete(T: Optional[RKDNode], x: List[float]) -> Optional[RKDNode]:
    if not T:
        return None
    i = T.discr
    if T.key == x:
        return delete_root(T)
    if x[i] < T.key[i]:
        T.left = delete(T.left, x)
    else:
        T.right = delete(T.right, x)
    update_size(T)
    return T

class RelaxedKDTree:
    def __init__(self, points: Optional[List[List[float]]] = None):
        self.root: Optional[RKDNode] = None
        if points:
            for p in points:
                self.insert(p)

    def insert(self, x: List[float]):
        self.root = insert(self.root, x)

    def delete(self, x: List[float]):
        self.root = delete(self.root, x)

    def partial_match(self, query: List[Union[float, str]]) -> tuple[int, int]:
        def _pm(node:RKDNode):
            nonlocal visits, matches
            if node is None:
                return
            visits += 1
            if all(q == '?' or q == v for q, v in zip(query, node.key)):
                matches += 1
            if node.left and (
                query[node.discr] == '?' or 
                query[node.discr] == node.key[node.discr]
            ):
                _pm(node.left)
            if node.right and (
                query[node.discr] == '?' or 
                query[node.discr] == node.key[node.discr]
            ):
                _pm(node.right)

        visits, matches = 0, 0
        _pm(self.root)
        return visits, matches

    def print_tree(self):
        self.root.print()

if __name__ == '__main__':
    tree = RelaxedKDTree()
    for _ in range(100):
        vec = [random.uniform(0, 1) for _ in range(2)]
        tree.insert(vec)
        if _ % 5 == 5-1:
            tree.print_tree()
            input('Press ENTER to continue')